user-operator/src/main/java/io/strimzi/operator/user/UserController.java

```diff
@@ -0,0 +1,306 @@
+    private final ControllerQueue workQueue;
+    private final List<UserControllerLoop> threadPool;
+
+    private final String watchedNamespace;
+    private final String secretPrefix;
+    private final long reconcileIntervalMs;
+    private final long operationTimeoutMs;
+
+    private final SharedIndexInformer<Secret> secretInformer;
+    private final SharedIndexInformer<KafkaUser> userInformer;
+
+    private final Timer periodicReconciliationTimer;
+
+    /**
+     * Creates the StrimziPodSet controller. The controller should normally exist once per operator for cluster-wide mode
+     * or once per namespace for namespaced mode.
+     *
+     * @param config          User Operator configuration
+     * @param client          Kubernetes client
+     * @param userOperator    The User Operator which encapsulates the logic for updating the users
+     * @param metricsProvider Metrics provider for handling metrics
+     */
+    public UserController(UserOperatorConfig config, KubernetesClient client, KafkaUserOperator userOperator, MetricsProvider metricsProvider) {
+        this.userOperator = userOperator;
+
+        // Store some useful settings into local fields
+        this.watchedNamespace = config.getNamespace();
+        this.secretPrefix = config.getSecretPrefix();
+        this.reconcileIntervalMs = config.getReconciliationIntervalMs();
+        this.operationTimeoutMs = config.getOperationTimeoutMs();
+
+        // User selector is used to select the KafkaUser resources
+        Map<String, String> userSelector = (config.getLabels() == null || config.getLabels().toMap().isEmpty()) ? Map.of() : config.getLabels().toMap();
+
+        // Selector for the secrets contains the original KafkaUSer selector adn the Strimzi Kind label
+        Map<String, String> secretSelector = new HashMap<>(userSelector.size() + 1);
+        secretSelector.putAll(userSelector);
+        secretSelector.put("strimzi.io/kind", RESOURCE_KIND);
+
+        // Set up the metrics holder
+        this.metrics = new ControllerMetricsHolder(RESOURCE_KIND, Labels.fromMap(userSelector), metricsProvider);
+
+        // Set up the work queue
+        this.workQueue = new ControllerQueue(config.getWorkQueueSize(), this.metrics);
+
+        // Secret informer and lister is used to get events about Secrets and get Secrets quickly
+        this.secretInformer = client.secrets().inNamespace(watchedNamespace).withLabels(secretSelector).inform();
+        Lister<Secret> secretLister = new Lister<>(secretInformer.getIndexer());
+
+        // KafkaUser informer and lister is used to get events about Users and get Users quickly
+        this.userInformer = Crds.kafkaUserOperation(client).inNamespace(watchedNamespace).withLabels(userSelector).inform();
+        Lister<KafkaUser> userLister = new Lister<>(userInformer.getIndexer());
+
+        // Creates the timer used for the periodical reconciliation
+        periodicReconciliationTimer = new Timer();
```

Comment:
Again using a scheduled executor here would be a bit more modern, and you could also share it with each of the `UserControllerLoop` created below. So 51 threads down to 1 thread  for all timed things (or 2 threads if you just shared an executor for the ProgressWarning threads, and kept a separate one for periodic reconciliations).